### PR TITLE
Feature onBoarding flow milestone 3

### PR DESCRIPTION
**Pull request by mejdoo**
This PR introduces milestone 3 for the feature "Linear onBoarding flow"

Goals for milestone 3:

Fetching the SSO code from the custom backend : https://wearezeta.atlassian.net/browse/AN-6635
Starting the SSO flow automatically if the code is present : https://wearezeta.atlassian.net/browse/AN-6634
#### APK
[Download build #1236](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1236/artifact/build/artifact/wire-dev-PR2607-1236.apk)
[Download build #1237](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1237/artifact/build/artifact/wire-dev-PR2607-1237.apk)
[Download build #1238](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1238/artifact/build/artifact/wire-dev-PR2607-1238.apk)
[Download build #1239](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1239/artifact/build/artifact/wire-dev-PR2607-1239.apk)
[Download build #1240](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1240/artifact/build/artifact/wire-dev-PR2607-1240.apk)
[Download build #1262](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1262/artifact/build/artifact/wire-dev-PR2607-1262.apk)

### CODE REVIEW

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala [hunk 1]
```diff
@@ -86,6 +86,34 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
...
+
+  protected def fetchTokenAndStartSsoFlow(): Unit =
+    userAccountsController.ssoToken.head.foreach {
+      case Some(token) => verifySsoCode(token)
+      case None =>
+        ssoService.fetchSSO().flatMap {
+          case Right(SSOFound(ssoCode)) => {
+            ssoService.extractUUID(s"wire-$ssoCode").fold(Future.successful(())) { token =>
+              onVerifyingToken(true)
+              ssoService.verifyToken(token).flatMap { result =>
```
Comment (BradleyWire) [inline app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: Can we extract this to a private method to be a little more readable?

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala [hunk 2]
```diff
@@ -152,9 +179,9 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
     case error => showInlineSsoError(getString(R.string.sso_signin_error_try_again_message, error.code.toString))
   }
 
-  private def dismissSsoDialog() = getSsoDialog.foreach(_.dismiss())
+  private def dismissSsoDialog() =    if (getSsoDialog.isDefined) getSsoDialog.foreach(_.dismiss())
```
Comment (BradleyWire) [inline app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: Formatting  between the `=` and` if `

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala (hunk 1 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: done

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala [hunk 3]
```diff
@@ -152,9 +178,14 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
     case error => showInlineSsoError(getString(R.string.sso_signin_error_try_again_message, error.code.toString))
   }
 
-  private def dismissSsoDialog() = getSsoDialog.foreach(_.dismiss())
+  private def dismissSsoDialog() = if (getSsoDialog.isDefined) getSsoDialog.foreach(_.dismiss())
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: `forEach` already guarantees that the item is defined. it does not do anything if the item is `None`. why did you need to add this?

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala [hunk 4]
```diff
@@ -152,9 +178,14 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
...
 
-  private def showInlineSsoError(errorText: String) = Future.successful(getSsoDialog.foreach(_.setError(errorText)))
+  private def showInlineSsoError(errorText: String) = Future.successful(if (getSsoDialog.isDefined) getSsoDialog.foreach(_.setError(errorText)))
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: same here.

File: wire-android-sync-engine/zmessaging/src/main/scala/com/waz/model2/transport/responses/FetchSsoResponse.scala [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import com.waz.utils.JsonDecoder
+import org.json.JSONObject
+
+sealed class FetchSsoResponse
+object SSONotFound extends FetchSsoResponse
+case class SSOFound(sso: String) extends FetchSsoResponse
+
+object FetchSsoResponse {
+
+  implicit val fetchSsoResponseResponseDecoder: JsonDecoder[FetchSsoResponse] = new JsonDecoder[FetchSsoResponse] {
```
Comment (gizemb) [inline wire-android-sync-engine/zmessaging/src/main/scala/com/waz/model2/transport/responses/FetchSsoResponse.scala]: i think one "response" is enough for the naming.

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala [hunk 6]
```diff
@@ -86,6 +86,34 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
 
   protected def showSsoByDefault = false
 
+  private lazy val showSsoDialogFuture = Future.successful(extractTokenAndShowSSODialog(true))
+
+  protected def fetchSsoToken(): Unit =
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: we can define these methods directly in CustomBackendLoginFragment, if no other descendant of SSOFragment is using it.

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala (hunk 3 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: alright removed! I thought it's required to make sure that the dialog is defined before calling dismiss.

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala (hunk 6 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: done

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala [hunk 7]
```diff
@@ -86,6 +86,9 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
 
   protected def showSsoByDefault = false
 
+  protected lazy val showSsoDialogFuture = Future.successful(extractTokenAndShowSSODialog(true))
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: this can also be moved to CustomBackendLoginFragment i guess.

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala [hunk 8]
```diff
@@ -156,6 +157,11 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
 
   private def showInlineSsoError(errorText: String) = Future.successful(getSsoDialog.foreach(_.setError(errorText)))
 
+  protected def showSsoWebView(token: String) = {
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: does this need to be protected?

File: app/src/main/scala/com/waz/zclient/appentry/CustomBackendLoginFragment.scala [hunk 9]
```diff
@@ -56,6 +63,36 @@ class CustomBackendLoginFragment extends SSOFragment {
     super.onPause()
     activity.getWindow.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
   }
+  private def fetchSsoToken(): Unit =
```
Comment (gizemb) [inline app/src/main/scala/com/waz/zclient/appentry/CustomBackendLoginFragment.scala]: put one line of space between methods please :)

File: app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala (hunk 8 above)
Comment (mejdoo) [reply app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala]: yes in order to access to it from CustomBackendLoginFragment

File: wire-android-sync-engine/zmessaging/src/main/scala/com/waz/model2/transport/responses/FetchSsoResponse.scala [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+sealed class FetchSsoResponse
+object SSONotFound extends FetchSsoResponse
+case class SSOFound(sso: String) extends FetchSsoResponse
+
+object FetchSsoResponse {
+
+  implicit val fetchSsoResponseDecoder: JsonDecoder[FetchSsoResponse] = new JsonDecoder[FetchSsoResponse] {
+    override def apply(implicit js: JSONObject): FetchSsoResponse = {
+      import JsonDecoder._
+      val configUrl = decodeOptString('default_sso_code)
```
Comment (makingthematrix) [inline wire-android-sync-engine/zmessaging/src/main/scala/com/waz/model2/transport/responses/FetchSsoResponse.scala]: cosmetic: You can use `Option.fold` here:
```
decodeOptString('default_sso_code).fold(SSONotFound)(SSOFound(_))
```

File: app/src/main/scala/com/waz/zclient/appentry/CustomBackendLoginFragment.scala [hunk 11]
```diff
@@ -2,17 +2,26 @@ package com.waz.zclient.appentry
...
+
+import scala.concurrent.Future
 
 class CustomBackendLoginFragment extends SSOFragment {
 
   val onEmailLoginClick = EventStream[Unit]
 
   private lazy val backendController = inject[BackendController]
 
+  protected lazy val showSsoDialogFuture = Future.successful(extractTokenAndShowSSODialog(true))
```
Comment (makingthematrix) [inline app/src/main/scala/com/waz/zclient/appentry/CustomBackendLoginFragment.scala]: Inline calls to this instead of creating a `lazy val`. Or change it to `def`.

It's possible that this will not cause a bug in this case, but in general, this way of using Scala futures is unsafe. A future is "eager", i.e. it will start executing the moment it's created. If you create a future, assign it to a reference, and then use that reference twice, the future will be executed only once. 

Check this snippet: https://scastie.scala-lang.org/s6mZgvw4TkKt2a5uO0AfEw (Click "Save" to run it)
